File: backend/models/Question.js
import mongoose from 'mongoose';

const questionSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Title is required'],
    trim: true,
  },
  company: {
    type: String,
    required: [true, 'Company is required'],
    trim: true,
  },
  topic: {
    type: String,
    required: [true, 'Topic is required'],
    trim: true,
  },
  difficulty: {
    type: String,
    enum: ['Easy', 'Medium', 'Hard'],
    required: [true, 'Difficulty is required'],
  },
  type: {
    type: String,
    enum: ['MCQ', 'CODING'],
    required: [true, 'Type is required'],
  },
  // MCQ-specific fields
  questionText: {
    type: String,
    default: null,
  },
  options: {
    type: [String],
    default: [],
  },
  correctAnswer: {
    type: String,
    default: null,
  },
  explanation: {
    type: String,
    default: null,
  },
  // Detailed description of the problem
  description: {
    type: String,
    default: '',
  },
  // Additional company tags (question may be asked by multiple companies)
  companyTags: {
    type: [String],
    default: [],
  },
  // CODING-specific fields
  platform: {
    type: String,
    enum: ['LeetCode', 'GFG', 'CodeChef', null],
    default: null,
  },
  // LeetCode or other platform URL — also accessible as 'leetcodeLink'
  link: {
    type: String,
    default: null,
  },
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true },
});

// Index for common query patterns
questionSchema.index({ company: 1, topic: 1, difficulty: 1, type: 1 });

const Question = mongoose.model('Question', questionSchema);
export default Question;

File: backend-spring/src/main/java/com/learn2hire/model/Question.java
package com.learn2hire.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "questions")
@CompoundIndex(name = "company_topic_difficulty_type", def = "{'company': 1, 'topic': 1, 'difficulty': 1, 'type': 1}")
public class Question {

    @Id
    @JsonProperty("_id")
    private String id;

    private String title;

    private String company;

    private String topic;

    private String difficulty; // Easy, Medium, Hard

    private String type; // MCQ, CODING

    // MCQ-specific fields
    private String questionText;

    @Builder.Default
    private List<String> options = new ArrayList<>();

    private String correctAnswer;

    private String explanation;

    @Builder.Default
    private String description = "";

    @Builder.Default
    private List<String> companyTags = new ArrayList<>();

    // CODING-specific fields
    private String platform; // LeetCode, GFG, CodeChef

    private String link;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}
